src/domain/core/src/entities/metadata_chain.rs

```diff
@@ -146,6 +147,97 @@ pub trait MetadataChainExt: MetadataChain {
     fn iter_blocks_ref<'a>(&'a self, head: &'a BlockRef) -> DynMetadataStream<'a> {
         self.iter_blocks_interval_ref(head, None)
     }
+
+    async fn accept<E>(
+        &self,
+        visitors: &mut [&mut dyn MetadataChainVisitor<Error = E>],
+    ) -> Result<(), E>
+    where
+        E: Error + From<IterBlocksError>,
+    {
+        self.accept_by_ref(visitors, &BlockRef::Head).await
+    }
+
+    async fn accept_by_hash<E>(
+        &self,
+        visitors: &mut [&mut dyn MetadataChainVisitor<Error = E>],
+        head_hash: &Multihash,
+    ) -> Result<(), E>
+    where
+        E: Error + From<IterBlocksError>,
+    {
+        let mut decisions = vec![Decision::Next; visitors.len()];
+
+        self.accept_by_hash_with_decisions(&mut decisions, visitors, head_hash)
+            .await
+    }
+
+    async fn accept_by_ref<E>(
+        &self,
+        visitors: &mut [&mut dyn MetadataChainVisitor<Error = E>],
+        head: &BlockRef,
+    ) -> Result<(), E>
+    where
+        E: Error + From<IterBlocksError>,
+    {
+        let head_hash = self.resolve_ref(head).await.map_err(Into::into)?;
+
+        self.accept_by_hash(visitors, &head_hash).await
+    }
+
+    async fn accept_by_hash_with_decisions<E>(
+        &self,
+        decisions: &mut [Decision],
+        visitors: &mut [&mut dyn MetadataChainVisitor<Error = E>],
```

Comment:
Since these two are so interconnected logically, shouldn't it be wrapped in a struct, something like `VisitingState` ?

Reply:
If you decide to add a counter of stopped visitors, it would logically belong to this struct very naturally.

Reply:
Moved to https://github.com/kamu-data/kamu-cli/pull/543#discussion_r1517884935